Whitelisting Onelake API & Workspace PL FQDNs

SmritiAgrawal04 opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

# Rationale for this change
 
1. With the release of the [WS PL FQDNs](https://learn.microsoft.com/en-us/fabric/security/security-workspace-level-private-links-overview#connecting-to-workspaces) with workspace private links, we need to add support for the new workspace FQDNs (“{wsid}.z{xy}.dfs.fabric.microsoft.com), so users can continue to use Delta-RS with OneLake and workspace private links.
2. We have also added support for "-api.onelake" FQDNs. as well [Ref](https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api#additional-onelake-endpoints).

<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?
1. Valid WS-PL FQDN egs:
* https://c047b3e34e89407a98d7cf9949ae92a3.zc0.onelake.fabric.microsoft.com
* https://c047b3e34e89407a98d7cf9949ae92a3.zc0.dfs.fabric.microsoft.com
* https://c047b3e34e89407a98d7cf9949ae92a3.zc0.blob.fabric.microsoft.com
3. Invalid WS-PL FQDN egs:
* https://c047b3e34e89407a98d7cf9949ae92a3.z12.blob.fabric.microsoft.com // xy mismatch 
* https://c047b3e34e89407a98d7cf9949ae923.z12.blob.fabric.microsoft.com  // workspaceid 31 chars
5. New API (regional & non-regional) endpoint support
* https://uksouth-api.onelake.fabric.microsoft.com
* https://api.onelake.fabric.microsoft.com

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?
Yes. The documentaion is as below:
1. https://learn.microsoft.com/en-us/fabric/security/security-workspace-level-private-links-overview#connecting-to-workspaces
7. https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api#additional-onelake-endpoints

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please call them out.
-->

SmritiAgrawal04 commented on src/azure/builder.rs:
```diff
@@ -1184,10 +1253,34 @@ mod tests {
         let mut builder = MicrosoftAzureBuilder::new();
         builder
-            .parse_url("https://account.blob.fabric.microsoft.com/container")
+            .parse_url("https://account.blob.fabric.microsoft.com/")
```
By mistake. Reverted.

alamb commented on src/azure/builder.rs:
```diff
@@ -671,36 +673,86 @@ impl MicrosoftAzureBuilder {
                 }
-                Some((a, "dfs.fabric.microsoft.com")) | Some((a, "blob.fabric.microsoft.com")) => {
-                    self.account_name = Some(validate(a)?);
-                    // Attempt to infer the container name from the URL
```
why remove this comment? It seems helpful

alamb commented on src/azure/builder.rs:
```diff
@@ -671,36 +673,86 @@ impl MicrosoftAzureBuilder {
+                    self.container_name = Some(validate(parsed.username())?);
+                    self.account_name = Some(validate(a)?);
+                    self.use_fabric_endpoint = true.into();
+                } else if let Some(a) = host.strip_suffix("-api.onelake.fabric.microsoft.com") {
```
- Is *-api.onelake.fabric.microsoft.com a publicly documented endpoint? If yes, can you point to the Microsoft doc so we can cite it in code/tests?

I don't see it in https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api

alamb reviewed (commented):
Thanks @SmritiAgrawal04  -- i left some more questions


BTW I asked codex for a review of this PR like this

```text
please review this PR (checked out in this directory): https://github.com/apache/arrow-rs-object-store/pull/552

  Specifically ensure that the format of the URLs supported matches publically published microsoft urls and provide documentation references
```

And this was the output which I found helpful

───────────────────────────────────────────────────────────────────────────────────────────────

• Findings

  - High: -api.onelake.fabric.microsoft.com is accepted for both abfs(s) and https, but I can’t find this FQDN in
    the public OneLake URL docs. The published endpoints are onelake.dfs.fabric.microsoft.com and workspace-scoped
    https://{workspaceId}.z{xy}.dfs.fabric.microsoft.com/... (see https://learn.microsoft.com/en-us/fabric/onelake/
    onelake-access-api). If -api.onelake is internal or not publicly documented, this should be gated or removed.
  - Medium: The WS‑PL regex no longer verifies that z{xy} matches the first two hex chars of the workspace id (src/
    azure/builder.rs:692-708). The published WS‑PL format is https://
    {workspaceId}.z{first2}.dfs.fabric.microsoft.com (same doc). Without the check, non‑published FQDNs are
    accepted.
  - Medium: WS‑PL regex allows .onelake.fabric.microsoft.com (src/azure/builder.rs:692-699). The public docs list
    dfs (and blob endpoints for blob access), but I don’t see onelake as a WS‑PL subdomain. Please confirm with a
    Microsoft reference or remove to match published formats.
  - Low: WS‑PL regex only allows lowercase hex ([0-9a-f]{32}) (src/azure/builder.rs:697). If Microsoft treats
    workspace IDs as case‑insensitive GUIDs in URLs, this will reject valid uppercase forms; consider documenting
    that only lowercase is supported or make the regex case‑insensitive.

  Doc references

  - OneLake access API URL formats: https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api
  - Private Link / workspace-specific OneLake endpoints (if this PR targets WS‑PL): please confirm the exact doc
    section that defines the workspaceId.z{xy}.dfs.fabric.microsoft.com pattern and whether any onelake WS‑PL host
    is documented (I could not find it in public docs).

  Questions / assumptions

  - Is *-api.onelake.fabric.microsoft.com a publicly documented endpoint? If yes, can you point to the Microsoft
    doc so we can cite it in code/tests?
  - Should WS‑PL accept only dfs/blob subdomains, or is .onelake.fabric.microsoft.com explicitly published?

SmritiAgrawal04 commented on src/azure/builder.rs:
```diff
@@ -671,36 +673,86 @@ impl MicrosoftAzureBuilder {
-                    }
+            "https" => {
+                // Regex to match WS-PL FQDN:
+                // "{workspaceid}.z??.(onelake|dfs|blob).fabric.microsoft.com"
```
Added for WS-PL DFS/ Blob endpoints. We are waiting for PM to confirm on ABFSS & WS-PL onelake domains.

SmritiAgrawal04 commented:
> Thanks @SmritiAgrawal04 -- i left some more questions
> 
> BTW I asked codex for a review of this PR like this
> 
> ```
> please review this PR (checked out in this directory): https://github.com/apache/arrow-rs-object-store/pull/552
> 
>   Specifically ensure that the format of the URLs supported matches publically published microsoft urls and provide documentation references
> ```
> 
> And this was the output which I found helpful
> 
> ───────────────────────────────────────────────────────────────────────────────────────────────
> 
> • Findings
> 
> * High: -api.onelake.fabric.microsoft.com is accepted for both abfs(s) and https, but I can’t find this FQDN in
>   the public OneLake URL docs. The published endpoints are onelake.dfs.fabric.microsoft.com and workspace-scoped
>   https://{workspaceId}.z{xy}.dfs.fabric.microsoft.com/... (see https://learn.microsoft.com/en-us/fabric/onelake/
>   onelake-access-api). If -api.onelake is internal or not publicly documented, this should be gated or removed.
> * Medium: The WS‑PL regex no longer verifies that z{xy} matches the first two hex chars of the workspace id (src/
>   azure/builder.rs:692-708). The published WS‑PL format is https://
>   {workspaceId}.z{first2}.dfs.fabric.microsoft.com (same doc). Without the check, non‑published FQDNs are
>   accepted.
> * Medium: WS‑PL regex allows .onelake.fabric.microsoft.com (src/azure/builder.rs:692-699). The public docs list
>   dfs (and blob endpoints for blob access), but I don’t see onelake as a WS‑PL subdomain. Please confirm with a
>   Microsoft reference or remove to match published formats.
> * Low: WS‑PL regex only allows lowercase hex ([0-9a-f]{32}) (src/azure/builder.rs:697). If Microsoft treats
>   workspace IDs as case‑insensitive GUIDs in URLs, this will reject valid uppercase forms; consider documenting
>   that only lowercase is supported or make the regex case‑insensitive.
> 
> Doc references
> 
> * OneLake access API URL formats: https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api
> * Private Link / workspace-specific OneLake endpoints (if this PR targets WS‑PL): please confirm the exact doc
>   section that defines the workspaceId.z{xy}.dfs.fabric.microsoft.com pattern and whether any onelake WS‑PL host
>   is documented (I could not find it in public docs).
> 
> Questions / assumptions
> 
> * Is *-api.onelake.fabric.microsoft.com a publicly documented endpoint? If yes, can you point to the Microsoft
>   doc so we can cite it in code/tests?
> * Should WS‑PL accept only dfs/blob subdomains, or is .onelake.fabric.microsoft.com explicitly published?

Hi @alamb,

I have addressed all comments. About, finding #1 & #3, we plan to add it to the public documentation, the PR for which is already out. I request to approve these changes meanwhile. Thanks

alamb commented on src/azure/builder.rs:
```diff
@@ -664,48 +666,99 @@ impl MicrosoftAzureBuilder {
+                
+                let ws_pl_regex = WS_PL_REGEX.get_or_init(|| {
+                    Regex::new(
+                        r"(?i)^(?P<workspaceid>[0-9a-f]{32})\.z(?P<xy>[0-9a-f]{2})\.(onelake|dfs|blob)\.fabric\.microsoft\.com$"
```
What is the rationale for using regexp math here rather than just hostname splitting? The examples / tests you have added don't seem to test for invalid URLs / that don't follow a simple <container>.<account>... format

SmritiAgrawal04 replied:
1. Workspace private link FQDNs follow a pattern unlike other FQDNs that had been whitelisted earlier. WS‑PL hostnames have a very specific, structured format with invariants that we want to validate, not merely parse. If you verify the FQDN syntax [here ](https://learn.microsoft.com/en-us/fabric/security/security-workspace-level-private-links-overview#connecting-to-workspaces), you'd note that a simple split('.') can tell “there are N labels” and extract strings, but it doesn’t naturally enforce:

* wsid is exactly 32 hex chars,
* xy is exactly 2 hex chars,
* and xy == wsid[0..2].

2. Added the test cases for invalid ws-pl fqdn.

alamb commented:
> I have addressed all comments. About, finding https://github.com/apache/arrow-rs-object-store/issues/1 & https://github.com/apache/arrow-rs-object-store/pull/3, we plan to add it to the public documentation, the PR for which is already out. I request to approve these changes meanwhile. Thanks

Some comments remain unaddressed. In terms of documentation, please include it in this PR
